www/artworks/post.php

```diff
@@ -0,0 +1,139 @@
+<? /** @noinspection PhpComposerExtensionStubsInspection, PhpIncludeInspection */
+
+use Ramsey\Uuid\Uuid;
+
+require_once('Core.php');
+
+/**
+ * @return array<string> an array containing [0] the path to the uploaded image and [1] the path to the thumbnail
+ * @throws \Exceptions\InvalidImageUploadException
+ */
+function handleImageUpload($uploadTmp): array {
+	$uploadInfo = getimagesize($uploadTmp);
+
+	if ($uploadInfo === false) {
+		throw new Exceptions\InvalidImageUploadException();
+	}
+
+	if ($uploadInfo[2] !== IMAGETYPE_GIF && $uploadInfo[2] !== IMAGETYPE_JPEG && $uploadInfo[2] !== IMAGETYPE_PNG) {
+		throw new Exceptions\InvalidImageUploadException();
+	}
+
+	$uid = Uuid::uuid4()->toString();
+	$ext = image_type_to_extension($uploadInfo[2]);
+
+	$uploadPath = COVER_ART_UPLOAD_PATH . $uid . $ext;
+	$thumbPath = COVER_ART_UPLOAD_PATH . $uid . '.thumb' . $ext;
```

Comment:
It's worth taking a moment to ensure we like this naming convention. We end up with files that look like this:

```
$ ls -1 www/images/uploads/
02584f01-fc89-4dec-aeff-9db5f8b97055.jpeg
02584f01-fc89-4dec-aeff-9db5f8b97055.thumb.jpeg
0ee63a18-eda7-4289-9a95-d5e6742ec08e.jpeg
0ee63a18-eda7-4289-9a95-d5e6742ec08e.thumb.jpeg
361a0483-940f-4520-b09a-d36f89c937c0.jpeg
361a0483-940f-4520-b09a-d36f89c937c0.thumb.jpeg
97063843-0b29-48d8-a215-14898fbf1dff.jpeg
97063843-0b29-48d8-a215-14898fbf1dff.thumb.jpeg
c3803beb-a3b6-4e18-977b-8c3ab6b9bec8.jpeg
c3803beb-a3b6-4e18-977b-8c3ab6b9bec8.thumb.jpeg
cfa67358-40eb-44d2-abef-5b67bbea1e77.jpeg
cfa67358-40eb-44d2-abef-5b67bbea1e77.thumb.jpeg
ea2b1c23-df11-49ef-bdc7-00d0a17ca353.jpeg
ea2b1c23-df11-49ef-bdc7-00d0a17ca353.thumb.jpeg```

Alternatives

1. A `thumbs/` subdirectory
2. A different separator, e.g., `_thumb.jpeg

The reason I ask is that we'll need to add a new property to the `Artwork` class:

```
_ThumbnailUrl
```

that is derived from `ImageFilesystemPath`. We might want to make it easy on ourselves to derive the thumbnail URLs. Maybe the dot will work, but we'll have to careful with the string or regex logic.

Reply:
GUIDs are not very ergonomic in the case where we have to go in and examine or manipulate files by hand. Instead we can simply name the file after the DB primary key value (i.e. `1.jpg`, `2.jpg`, etc.). That way we also avoid needing another column for GUID. Also let's go with `.jpg` and not `.jpeg`.

Reply:
If the filename is going derived from the PK we can remove the column that stores the file location, right? As the file location will always be `/www/images/uploads/{id}.jpg`.

Reply:
@acabal, how do feel about the full-resolution image URLs being guessable? Going with `1.jpg`, `2.jpg`, etc. would allow potentially annoying scrapers:

```
for i in $(seq 1 500); do
  curl -O https://standardebooks.org/www/images/uploads/$i.jpg
done
```

(This came up in a [thread](https://groups.google.com/g/standardebooks/c/KrXyzZVZig4/m/E7BRuJ2wBQAJ) in Mar–Apr 2021 about bulk downloading ebooks.)

If you don't think that's too big of an issue, or it's still worth the trade-off, then I agree with you and Job that we can simplify here and remove the `ImageFilesystemPath` column. We'll still add these to properties to the `Artwork` class,

```
_Url
_ThumbnailUrl
```

but they will be trivial.

Reply:
Having them guessable is fine. Unlikely to be scraped I think, nobody is seriously scraping hi-res public domain oil paintings without any accompanying metadata. If it becomes a problem we can address it at that time.

Reply:
Cool, I've implemented this change so that uploads are stored at `/images/uploads/{id}.jpg` and thumbs at `/images/uploads/{id}.thumb.jpg`.

Another potential risk of is that if upload URLs are predictable, it would be possible to use the SE website as a free image hosting service since even unapproved uploads will be available. This also seems pretty unlikely (and not too difficult to detect) but maybe worth being aware of.